### PR TITLE
PP-2417 Accepting Maestro

### PR DESCRIPTION
**Pull request by rauligs**
## WHAT
-  Add by default 'non 3ds required' card types as *false.* So Created a new column to identify cards requiring _3ds_ to be enabled in order to be used.
- Add 'Maestro' as a new card type.
- Add 'requires_3ds' in card type _Json_ response.
- Handle update gateway account accepted card types 3ds mismatch configuration (409 Conflict).
- Handle _toggle 3ds_ mismatch configuration (having card types with 3ds required), returns 409 Conflict.
- Check 3ds required on Card Auth. So if gateway account has _3ds disabled_ but the card type requires 3ds will throw an exception and set the charge to a final status as `AUTHORISATION ABORTED`.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/model/domain/GatewayAccountEntity.java [hunk 1]
```diff
@@ -214,4 +215,10 @@ public void setType(Type type) {
     public boolean hasEmailNotificationsEnabled() {
         return emailNotification != null && emailNotification.isEnabled();
     }
+
+    public boolean hasAnyAcceptedCardType3dsRequired() {
+        return cardTypes.stream()
+                .filter(CardTypeEntity::isRequires3ds)
+                .count() > 0;
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/model/domain/GatewayAccountEntity.java]: `return cardTypes.stream().anyMatch(CardTypeEntity::isRequires3ds);`

Comment (rauligs) [reply src/main/java/uk/gov/pay/connector/model/domain/GatewayAccountEntity.java]: indeed!

File: src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java [hunk 2]
```diff
@@ -12,33 +13,55 @@
...
     @Transactional
-    public ChargeEntity preOperation(ChargeEntity chargeEntity) {
-        chargeEntity = preOperation(chargeEntity, OperationType.AUTHORISATION, getLegalStates(), AUTHORISATION_READY);
-        getPaymentProviderFor(chargeEntity).generateTransactionId().ifPresent(chargeEntity::setGatewayTransactionId);
+    public ChargeEntity preOperation(ChargeEntity chargeEntity, AuthCardDetails authCardDetails) {
+
+        String cardBrand = authCardDetails.getCardBrand();
+        ChargeEntity reloadedCharge = chargeDao.merge(chargeEntity);
+
+        if (!reloadedCharge.getGatewayAccount().isRequires3ds() && cardBrandRequires3ds(cardBrand)) {
```
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java:45]: :+1:

File: src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java [hunk 3]
```diff
@@ -12,33 +13,55 @@
...
+            chargeEntity = chargeDao.mergeAndNotifyStatusHasChanged(reloadedCharge, Optional.empty());
+        } else {
+            chargeEntity = preOperation(chargeEntity, OperationType.AUTHORISATION, getLegalStates(), AUTHORISATION_READY);
+            getPaymentProviderFor(chargeEntity).generateTransactionId().ifPresent(chargeEntity::setGatewayTransactionId);
+        }
+
         return chargeEntity;
     }
 
+    private boolean cardBrandRequires3ds(String cardBrand) {
```
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java:58]: We had a chat about weather we should also check that the card type is one of the accepted ones on the account and decided that for now it should be sufficient to only check that the account has 3DS on should the card require 3DS.

**Review by ckalista — DISMISSED** [review]
Looks good and ready to go as soon as we get a green build.

**Review by ckalista — APPROVED** [review]
Re-approving after rebase.

**Pull request merged**